File: app/lib/services/submission_service.dart
import '../models/submission.dart';
import '../config/api_config.dart';
import 'api_client.dart';

class SubmissionService {
  final ApiClient _apiClient = ApiClient();

  // Create new submission for a program
  Future<Submission> createSubmission(String programId, String title) async {
    try {
      final response = await _apiClient.post(
        '${ApiConfig.apiBase}/programs/$programId/submissions',
        {'title': title},
      );
      final data = _apiClient.parseResponse(response);

      return Submission.fromJson(data['submission'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to create submission: $e');
    }
  }

  // List submissions with optional filters
  Future<List<SubmissionListItem>> listSubmissions({
    String? programId,
    int limit = 50,
    int offset = 0,
  }) async {
    try {
      final queryParams = <String>[];
      if (programId != null) {
        queryParams.add('program_id=$programId');
      }
      queryParams.add('limit=$limit');
      queryParams.add('offset=$offset');

      final queryString = queryParams.isNotEmpty ? '?${queryParams.join('&')}' : '';
      final response = await _apiClient.get(
        '${ApiConfig.apiBase}/submissions$queryString',
      );
      final data = _apiClient.parseResponse(response);

      final submissionsList = data['submissions'] as List<dynamic>?;
      if (submissionsList == null) {
        return [];
      }

      final submissions = submissionsList
          .map((json) => SubmissionListItem.fromJson(json as Map<String, dynamic>))
          .toList();

      return submissions;
    } catch (e) {
      throw Exception('Failed to load submissions: $e');
    }
  }

  // Get single submission by ID
  Future<Submission> getSubmission(String id) async {
    try {
      final response = await _apiClient.get('${ApiConfig.apiBase}/submissions/$id');
      final data = _apiClient.parseResponse(response);

      return Submission.fromJson(data['submission'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to load submission: $e');
    }
  }

  // Get all messages for a submission
  Future<List<MessageWithAuthor>> getMessages(String submissionId) async {
    try {
      final response = await _apiClient.get(
        '${ApiConfig.apiBase}/submissions/$submissionId/messages',
      );
      final data = _apiClient.parseResponse(response);

      final messagesList = data['messages'] as List<dynamic>?;
      if (messagesList == null) {
        return [];
      }

      final messages = messagesList
          .map((json) => MessageWithAuthor.fromJson(json as Map<String, dynamic>))
          .toList();

      return messages;
    } catch (e) {
      throw Exception('Failed to load messages: $e');
    }
  }

  // Create new message in a submission
  Future<SubmissionMessage> createMessage(
    String submissionId,
    String content, {
    String? youtubeUrl,
  }) async {
    try {
      final payload = <String, dynamic>{
        'content': content,
      };
      if (youtubeUrl != null && youtubeUrl.isNotEmpty) {
        payload['youtube_url'] = youtubeUrl;
      }

      final response = await _apiClient.post(
        '${ApiConfig.apiBase}/submissions/$submissionId/messages',
        payload,
      );
      final data = _apiClient.parseResponse(response);

      return SubmissionMessage.fromJson(data['message'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to create message: $e');
    }
  }

  // Mark a message as read
  Future<void> markMessageAsRead(String messageId) async {
    try {
      final response = await _apiClient.put(
        '${ApiConfig.apiBase}/messages/$messageId/read',
        {},
      );
      _apiClient.parseResponse(response);
    } catch (e) {
      throw Exception('Failed to mark message as read: $e');
    }
  }

  // Get unread message counts
  Future<UnreadCounts> getUnreadCount({String? programId}) async {
    try {
      final queryString = programId != null ? '?program_id=$programId' : '';
      final response = await _apiClient.get(
        '${ApiConfig.apiBase}/submissions/unread-count$queryString',
      );
      final data = _apiClient.parseResponse(response);

      return UnreadCounts.fromJson(data as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to get unread count: $e');
    }
  }
}

File: backend/internal/services/submission_service.go
package services

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/xuangong/backend/internal/models"
	"github.com/xuangong/backend/internal/repositories"
	appErrors "github.com/xuangong/backend/pkg/errors"
	"github.com/xuangong/backend/pkg/youtube"
)

type SubmissionService struct {
	submissionRepo *repositories.SubmissionRepository
	programRepo    *repositories.ProgramRepository
}

func NewSubmissionService(submissionRepo *repositories.SubmissionRepository, programRepo *repositories.ProgramRepository) *SubmissionService {
	return &SubmissionService{
		submissionRepo: submissionRepo,
		programRepo:    programRepo,
	}
}

// CreateSubmission creates a new submission for a program
func (s *SubmissionService) CreateSubmission(ctx context.Context, programID, userID uuid.UUID, title string) (*models.Submission, error) {
	// Validate title
	if title == "" {
		return nil, appErrors.NewBadRequestError("Title cannot be empty")
	}

	// Verify program exists
	program, err := s.programRepo.GetByID(ctx, programID)
	if err != nil {
		return nil, appErrors.NewInternalError("Failed to verify program").WithError(err)
	}
	if program == nil {
		return nil, appErrors.NewNotFoundError("Program")
	}

	// Create submission
	submission, err := s.submissionRepo.Create(ctx, programID, userID, title)
	if err != nil {
		return nil, appErrors.NewInternalError("Failed to create submission").WithError(err)
	}

	return submission, nil
}

// GetSubmission retrieves a submission by ID with access control
func (s *SubmissionService) GetSubmission(ctx context.Context, id, userID uuid.UUID, isAdmin bool) (*models.Submission, error) {
	submission, err := s.submissionRepo.GetByID(ctx, id, userID, isAdmin)
	if err != nil {
		if errors.Is(err, repositories.ErrAccessDenied) {
			return nil, appErrors.NewAuthorizationError("You don't have access to this submission")
		}
		if errors.Is(err, repositories.ErrSubmissionNotFound) {
			return nil, appErrors.NewNotFoundError("Submission")
		}
		return nil, appErrors.NewInternalError("Failed to fetch submission").WithError(err)
	}

	return submission, nil
}

// ListSubmissions retrieves submissions with filters and access control
func (s *SubmissionService) ListSubmissions(ctx context.Context, programID *uuid.UUID, userID uuid.UUID, isAdmin bool, limit, offset int) ([]models.SubmissionListItem, error) {
	// Validate pagination
	if limit <= 0 || limit > 100 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}

	submissions, err := s.submissionRepo.List(ctx, programID, userID, isAdmin, limit, offset)
	if err != nil {
		return nil, appErrors.NewInternalError("Failed to list submissions").WithError(err)
	}

	return submissions, nil
}

// CreateMessage adds a message to a submission
func (s *SubmissionService) CreateMessage(ctx context.Context, submissionID, userID uuid.UUID, isAdmin bool, content string, youtubeURL *string) (*models.SubmissionMessage, error) {
	// Validate content
	if content == "" {
		return nil, appErrors.NewBadRequestError("Message content cannot be empty")
	}

	// Validate YouTube URL if provided
	if youtubeURL != nil && *youtubeURL != "" {
		if _, err := youtube.ValidateURL(*youtubeURL); err != nil {
			return nil, appErrors.NewBadRequestError(fmt.Sprintf("Invalid YouTube URL: %v", err))
		}
	}

	// Verify access to submission
	submission, err := s.submissionRepo.GetByID(ctx, submissionID, userID, isAdmin)
	if err != nil {
		if errors.Is(err, repositories.ErrAccessDenied) {
			return nil, appErrors.NewAuthorizationError("You don't have access to this submission")
		}
		if errors.Is(err, repositories.ErrSubmissionNotFound) {
			return nil, appErrors.NewNotFoundError("Submission")
		}
		return nil, appErrors.NewInternalError("Failed to verify submission access").WithError(err)
	}
	if submission == nil {
		return nil, appErrors.NewNotFoundError("Submission")
	}

	// Create message
	message, err := s.submissionRepo.CreateMessage(ctx, submissionID, userID, content, youtubeURL)
	if err != nil {
		return nil, appErrors.NewInternalError("Failed to create message").WithError(err)
	}

	return message, nil
}

// GetMessages retrieves all messages for a submission with access control
func (s *SubmissionService) GetMessages(ctx context.Context, submissionID, userID uuid.UUID, isAdmin bool) ([]models.MessageWithAuthor, error) {
	messages, err := s.submissionRepo.GetMessages(ctx, submissionID, userID, isAdmin)
	if err != nil {
		if errors.Is(err, repositories.ErrAccessDenied) {
			return nil, appErrors.NewAuthorizationError("You don't have access to this submission")
		}
		if errors.Is(err, repositories.ErrSubmissionNotFound) {
			return nil, appErrors.NewNotFoundError("Submission")
		}
		return nil, appErrors.NewInternalError("Failed to fetch messages").WithError(err)
	}

	return messages, nil
}

// MarkMessageAsRead marks a message as read by a user
func (s *SubmissionService) MarkMessageAsRead(ctx context.Context, userID, messageID uuid.UUID) error {
	err := s.submissionRepo.MarkMessageAsRead(ctx, userID, messageID)
	if err != nil {
		if errors.Is(err, repositories.ErrMessageNotFound) {
			return appErrors.NewNotFoundError("Message")
		}
		return appErrors.NewInternalError("Failed to mark message as read").WithError(err)
	}

	return nil
}

// GetUnreadCount returns unread message counts at various levels
func (s *SubmissionService) GetUnreadCount(ctx context.Context, userID uuid.UUID, programID *uuid.UUID) (*models.UnreadCounts, error) {
	counts, err := s.submissionRepo.GetUnreadCount(ctx, userID, programID)
	if err != nil {
		return nil, appErrors.NewInternalError("Failed to get unread counts").WithError(err)
	}

	return counts, nil
}

// SoftDeleteSubmission soft deletes a submission (admin only)
func (s *SubmissionService) SoftDeleteSubmission(ctx context.Context, id, userID uuid.UUID, isAdmin bool) error {
	// Only admins can delete
	if !isAdmin {
		return appErrors.NewAuthorizationError("Only admins can delete submissions")
	}

	err := s.submissionRepo.SoftDelete(ctx, id)
	if err != nil {
		if errors.Is(err, repositories.ErrAlreadyDeleted) {
			return appErrors.NewNotFoundError("Submission")
		}
		return appErrors.NewInternalError("Failed to delete submission").WithError(err)
	}

	return nil
}
